### PR TITLE
Ensure Equivalence With DPO Padding and Padding Free

### PR DESCRIPTION
**Pull request by fabianlim**
Actually I believe that padding-free (packing) is correct. 
- this PR fixes the non-packing case, where `concatenated_forward` assumes the `pad_token_id = 0` by mistake

### CODE REVIEW

File: open_instruct/dpo_tune_cache.py [hunk 1]
```diff
@@ -76,9 +76,6 @@
     simpo_loss,
     wpo_loss,
 )
-from open_instruct.padding_free_collator import (
```
Comment (garrett361) [inline open_instruct/dpo_tune_cache.py:79]: Intentionally deleted?

File: open_instruct/dpo_utils.py [hunk 2]
```diff
@@ -234,7 +237,11 @@ def concatenated_forward(
     We do this to avoid doing two forward passes, because it's faster for FSDP.
     """
     if not packing:
-        concatenated_batch = concatenated_inputs(batch)
+        try:
+            pad_token_id = model.tokenizer.pad_token_id
```
Comment (garrett361) [inline open_instruct/dpo_utils.py:241]: Do models usually have a `tokenizer` attr? I'm finding not, for bamba and granite. Unless the attr is set somewhere after init.

Comment (fabianlim) [reply open_instruct/dpo_utils.py:241]: they will usually have, though the model is typed as `torch.nn.Module` so I thought in the OI use-case it may not be gauranteed.

File: open_instruct/dpo_tune_cache.py (hunk 1 above)
Comment (fabianlim) [reply open_instruct/dpo_tune_cache.py:79]: yes there is a dup two lines belwo

File: open_instruct/padding_free_collator.py [hunk 3]
```diff
@@ -129,7 +131,7 @@ def concatenated_inputs(
 
     if "seq_idx" in chosen_features:
         ret[f"{tag}seq_idx"] = torch.cat(
-            [chosen_features["seq_idx"], rejected_features["seq_idx"] + chosen_features["seq_idx"][0, -1]], dim=-1
+            [chosen_features["seq_idx"], rejected_features["seq_idx"] + len(chosen_features["cu_seq_lens_k"])- 1], dim=-1
```
Comment (garrett361) [inline open_instruct/padding_free_collator.py:134]: I'm confused about why these two are not equivalent

Comment (fabianlim) [reply open_instruct/padding_free_collator.py:134]: > is it an off by one error? Seems like chosen_features["seq_idx"][0, -1] + 1 == len(chosen_features["cu_seq_lens_k"])- 1

Yes correct, so I thought its more obvious in the rewritten way, since folks might not be as familiar with `seq_idx`. It is more obvious that `cu_seq_lens_k` has length `batch_size + 1`?

Comment (garrett361) [reply open_instruct/padding_free_collator.py:134]: think either is fine! thanks for catching this, again

File: tests/test_padding_free.py [hunk 4]
```diff
@@ -195,3 +208,68 @@ def test_padding_free(self, model_name: str, loss_type: str) -> None:
...
+            for k in b:
+                if torch.is_tensor(b[k]):
+                    b[k] = b[k].cuda()
+
+        assert batch["chosen_input_ids"].shape[0] == 2
+        assert pf_batch["chosen_input_ids"].shape[0] == 1
+        chosen_logps, rejected_logps, _ = concatenated_forward(model, batch, average_log_prob=False)
+        pf_chosen_logos, pf_rejected_logps, _ = concatenated_forward(pf_model, pf_batch, average_log_prob=False, packing=True)
+
+        torch.testing.assert_close(chosen_logps, pf_chosen_logos, atol=1e-3, rtol=1e-3)
```
Comment (garrett361) [inline tests/test_padding_free.py:273]: Were these previously passing with the off-by-one error?

**Review by garrett361 — APPROVED** [review]
LGTM, just one question.
